woocommerce/woocommerce-ios
# Domain web checkout flow with cart creation and a barebone success view

Part of #8558 


## Description


This PR includes the navigation & backend from selecting a domain to web checkout after creating a shopping cart behind the scene. When web checkout is complete (the same way we observe the URL to the "thank-you- URL like in store creation), a barebone success view is shown with a CTA to go back to the domain settings.

### Networking & Yosemite layer changes

After some back-and-forth, I decided to implement the domain cart creation in `PaymentRemote` instead of `DomainRemote`. I also tried reusing `PaymentRemote.createCart` to create a cart with a domain, but the parameters are quite different from purchasing a WPCOM plan product in store creation. Also, cart creation API requests do not work when using the default `URLEncoding` for the POST request body. I was able to reproduce in Postman. `URLEncoding` is fine for a WPCOM plan product, but not for a domain product with additional parameters at the product level. Therefore, I updated `DotcomRequest` to allow a custom parameter encoding strategy and pass `JSONEncoding` just for domain cart creation.

`PaymentRemote.createCart` with a domain returns a `CartResponse` instead of no return value in `PaymentRemote.createCart` with a plan product because this is going to be used for domain credit redemption later. When redeeming a domain credit (a free domain is included for the first year for certain WPCOM plans), the cart response is passed to another API request `me/transactions` to complete the redemption process.

### UI/UX

A new SwiftUI view `DomainPurchaseSuccessView` was created to show a barebone UI when a domain purchase is complete. In `DomainSettingsCoordinator`, after a domain is selected, it internally creates a cart with the domain async and then shows web checkout on success. When the web checkout succeeds, the new success view is shown with a CTA to go back to the domain settings. Please note that error handling in various places is still WIP, we don't have a design either.

## Testing instructions


Prerequisite: a WC store with a WPCOM plan is required, make sure the proxy is disabled so that you can see the web checkout page

- Log in if needed, and continue with a WPCOM store
- Go to the Menu tab
- Tap on the settings CTA
- Tap the `Domain` row
- Tap `Add a domain` or `Search for a domain` depending on if there are existing domains --> domain selector with paid domains should be shown, and each domain should have price info (with the barebone UI)
- Feel free to search for other domains --> each domain should have price info
- Select a domain and tap `Continue` --> if there's an error, an error should be logged in the console. I've been seeing an error with certain domain names, I'll look more into it separately. if a domain can be added to the cart, it should navigate to a webview to check out the domain
- Complete the web checkout --> when the checkout is successful, it should navigate to a barebone success screen with the purchased domain
- Tap `Done` --> it should navigate back to the domain settings, and the new domain should be shown shortly

## Screenshots


<img src="https://user-images.githubusercontent.com/1945542/213382723-0fbed5e4-d36b-4c96-b0af-8f41472f6183.png" width="300" />

https://user-images.githubusercontent.com/1945542/213382730-fbe42f93-432f-4e88-9af2-046f3f59da9b.mov



---
